### PR TITLE
backend events

### LINKED ISSUE #4443: Backends create sensu-system namespace and its entity
## Feature Suggestion

Sensu backend entities should be visible in the Sensu API and web app. 

## Possible Implementation

On backend start:

- Check for `sensu-system` namespace
** Create if missing
- Check for existing entity in the `sensu-system` named after the backend's local hostname
** If missing, create a new entity with `entity_class` of `backend` and populate the entity system information
** If one exists, don't touch it

Backend uses its own entity when reporting cluster state errors.

## Context

See: #4442

### PR DESCRIPTION
**Pull request by fguimond**
* create system namespace
* create backend entity
* create new error for vault access errors
* generate events when vault not available
* unit tests

Signed-off-by: Francis Guimond <francis@sensu.io>

## What is this change?

On backend startup create the `sensu-system` namespace as well as a backend entity (if necessary). When the secrets provider returns an access error log a critical event against the backend entity. When it's accessible again log an ok event against the entity. There is a mechanism to prevent "spamming" the event bus with backend events, so backend events are generated no more than every 30 seconds, unless the status changes.

## Why is this change necessary?

Closes #4443 and #4261 

## Does your change need a Changelog entry?

Yes

## Do you need clarification on anything?

No

## Were there any complications while making this change?

No

## Have you reviewed and updated the documentation for this change? Is new documentation required?

@hillaryfraley I believe this will need to be documented.

## How did you verify this change?

Manual testing:
* Create a Vault container: 
```
docker run --cap-add=IPC_LOCK -e 'VAULT_DEV_ROOT_TOKEN_ID=myroot' \                                                                                                                     
    -e 'VAULT_DEV_LISTEN_ADDRESS=0.0.0.0:8200' -p 8200:8200 vault
```
* Navigate to `http://localhost:8200`, login using the `myroot` secret and click on `secrets`
* Create a new secret with the following information
** Path: `sensu/prom`; key: `http_url`; value: `http://localhost:9100/metrics`; hit Save
* Create a check using this secret (path has to be `secrets/data/sensu/prom#http_url`

In the web ui or using sensuctl you will see an OK event under the `sensu-system/backendname` entity.

## Is this change a patch?

No

### CODE REVIEW

File: backend/resource/backend.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	return nil
+}
+
+func (br *BackendResource) GenerateBackendEvent(component string, status uint32, output string) error {
+	if br.backendEntity == nil {
+		return errors.New("backend entity doesn't exist")
+	}
+
+	now := time.Now().Unix()
+	checkName := fmt.Sprintf("%s-%s", corev2.BackendCheckName, component)
```
Comment (echlebek) [inline backend/resource/backend.go:82]: I don't think prepending all the backend event check names with "backend" is necessary. We already have "backend" as the entity name, which is part of the key to retrieve an event.

File: backend/resource/backend.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+			ObjectMeta: corev2.ObjectMeta{
+				Name:      checkName,
+				Namespace: systemNamespaceName,
+			},
+			Issued:   now,
+			Executed: now,
+			Output:   output,
+			Status:   status,
+		},
+		ObjectMeta: corev2.ObjectMeta{},
```
Comment (echlebek) [inline backend/resource/backend.go:103]: The namespace should be defined, and labels/annotations should be set to empty slices. This gives a nicer user experience around filters and selectors.

File: backend/secrets/error.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package secrets
+
+import "fmt"
+
+type ErrInvalidSecretInfo string
+type ErrNoProviderDefined struct{}
```
Comment (echlebek) [inline backend/secrets/error.go:6]: For errors with an invariant message, it's a little better and a little more idiomatic to use

```
var ErrNoProviderDefined = errors.New("no provider defined")
```

These can be easily compared by value, for example, `if err == ErrNoProviderDefined { // handle error }`.

In your use of ErrNoProviderDefined, you've returned a pointer value, which would not compare as easily. Similar for ErrSecretsNotSupported.

File: go.mod [hunk 4]
```diff
@@ -75,6 +75,7 @@ require (
 	go.etcd.io/etcd/client/v3 v3.5.0
 	go.etcd.io/etcd/server/v3 v3.5.0
 	go.etcd.io/etcd/tests/v3 v3.5.0
+	go.uber.org/atomic v1.7.0
```
Comment (echlebek) [inline go.mod:78]: :thinking: not sure where this dependency comes from, does it persist if you run `go mod tidy`?

**Review by echlebek — APPROVED** [review]
Great work Francis! I look forward to making use of these new concepts. Just a few small comments.

File: backend/resource/backend.go (hunk 1 above)
Comment (fguimond) [reply backend/resource/backend.go:82]: Agreed, fixed.

File: backend/resource/backend.go (hunk 2 above)
Comment (fguimond) [reply backend/resource/backend.go:103]: Fixed 👍

File: backend/secrets/error.go (hunk 3 above)
Comment (fguimond) [reply backend/secrets/error.go:6]: Fixed.

**Comment by fguimond** [comment]
Created https://github.com/sensu/sensu-enterprise-go/pull/2040 to return proper error when the Vault is not accessible.

**Pull request merged**